addons/bus/static/src/js/websocket_bus.js

```diff
@@ -0,0 +1,172 @@
+/** @odoo-module **/
+
+import { browser } from "@web/core/browser/browser";
+
+/**
+ * Event Websocket bus used to bind events on websocket notifications
+ *
+ * triggers on env.bus:
+ * - window_focus : when the window focus change (true for focused, false for blur)
+ */
+export class WebsocketBus {
+    constructor(env, websocketService) {
+        this.env = env;
+        this.websocketService = websocketService;
+        this._channels = [];
+        this._hasSubscribed = false;
+        this._isOdooFocused = true;
+        this._userUpdatePeriod = 30000; // update user presence every 30s
+
+        // bus presence
+        this._lastPresenceTime = new Date().getTime();
+
+        browser.addEventListener('focus', this._onFocusChange.bind(this, { focus: true }));
+        browser.addEventListener('blur', this._onFocusChange.bind(this, { focus: false }));
+        browser.addEventListener('unload', this._onFocusChange.bind(this, { focus: false }));
+        browser.addEventListener('click', this._onPresence.bind(this));
+        browser.addEventListener('keydown', this._onPresence.bind(this));
```

Comment:
it looks like the presence code is mixed with the websocket code. I think you should add a `presence_service` that will simply monitor the click/keydown events, and perform all the presence-related tasks